contracts/src/CrossChainApplications/NativeTokenBridge/ERC20TokenSource.sol

```diff
@@ -0,0 +1,224 @@
+import {ITokenSource} from "./ITokenSource.sol";
+import {ITeleporterMessenger, TeleporterMessageInput, TeleporterFeeInfo} from "../../Teleporter/ITeleporterMessenger.sol";
+import {ITeleporterReceiver} from "../../Teleporter/ITeleporterReceiver.sol";
+import {SafeERC20TransferFrom} from "../../Teleporter/SafeERC20TransferFrom.sol";
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+contract ERC20TokenSource is
+    ITeleporterReceiver,
+    IERC20TokenSource,
+    ITokenSource,
+    ReentrancyGuard
+{
+    // The address where the burned transaction fees are credited.
+    address public constant BURNED_TX_FEES_ADDRESS =
+        0x0100000000000000000000000000000000000000;
+    uint256 public constant MINT_NATIVE_TOKENS_REQUIRED_GAS = 100_000;
+    // Used to keep track of tokens burned through transactions on the destination chain. They can
+    // be reported to this contract to burn an equivalent number of tokens on this chain.
+    uint256 public destinationBurnedTotal;
+    bytes32 public immutable destinationBlockchainID;
+    address public immutable nativeTokenDestinationAddress;
+    address public immutable erc20ContractAddress;
+
+    // Used for sending and receiving Teleporter messages.
+    ITeleporterMessenger public immutable teleporterMessenger;
+
+    constructor(
+        address teleporterMessengerAddress,
+        bytes32 destinationBlockchainID_,
+        address nativeTokenDestinationAddress_,
+        address erc20ContractAddress_
+    ) {
+        require(
+            teleporterMessengerAddress != address(0),
+            "ERC20TokenSource: zero TeleporterMessenger address"
+        );
+        teleporterMessenger = ITeleporterMessenger(teleporterMessengerAddress);
+
+        require(
+            destinationBlockchainID_ != bytes32(0),
+            "ERC20TokenSource: zero destination chain ID"
+        );
+        require(
+            destinationBlockchainID_ !=
+                IWarpMessenger(0x0200000000000000000000000000000000000005)
+                    .getBlockchainID(),
+            "ERC20TokenSource: cannot bridge with same blockchain"
+        );
+        destinationBlockchainID = destinationBlockchainID_;
+
+        require(
+            nativeTokenDestinationAddress_ != address(0),
+            "ERC20TokenSource: zero destination contract address"
+        );
+        nativeTokenDestinationAddress = nativeTokenDestinationAddress_;
+
+        require(
+            erc20ContractAddress_ != address(0),
+            "ERC20TokenSource: invalid ERC20 contract address"
+        );
+        erc20ContractAddress = erc20ContractAddress_;
+    }
+
+    /**
+     * @dev See {ITeleporterReceiver-receiveTeleporterMessage}.
+     *
+     * Receives a Teleporter message and routes to the appropriate internal function call.
+     */
+    function receiveTeleporterMessage(
```

Comment:
Calling out that we should use `TeleporterUpgradable` here so that we can use the `onlyAllowedTeleporter` modifier. If we want to do this in a later PR, lets create an issue for it now. 👍

Reply:
I implemented it in this commit, but reverted the changes because the e2e setup doesn't currently deploy the registry, and I didn't want to go fooling around in there until the `general-e2e` PR is merged. 

Should we have a general ticket for all the cross-chain applications to use the registry?

Reply:
Seconding this. I'm in favor of creating a followup ticket to add registry and upgrade support to this set of contracts so that we don't further bloat this PR's scope.

Reply:
agree with above, lets create a new issue for it and defer